pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -106,14 +106,24 @@ func NewGarbageCollector(
 		sharedInformers:  sharedInformers,
 		ignoredResources: ignoredResources,
 	}
-	if err := gb.monitorsForResources(deletableResources); err != nil {
-		return nil, err
+	if err := gb.syncMonitors(deletableResources); err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to sync all monitors: %v", err))
 	}
 	gc.dependencyGraphBuilder = gb
 
 	return gc, nil
 }
 
+// SyncResourceMonitors starts or stops resource monitors as needed to ensure
+// that all (and only) those resources present in the map are monitored.
+func (gc *GarbageCollector) SyncResourceMonitors(deletableResources map[schema.GroupVersionResource]struct{}) error {
```

Comment:
Quoting https://github.com/kubernetes/kubernetes/issues/28713#issuecomment-234826276:

>I find the cause of the flake and will send a patch later.
>
>The failed test sets up a RC and a pod which depends on the RC, then it starts the GC, deletes the RC, and expects the GC to garbage collect the pod. It fails because the GC fails to observe the deletion of the RC. The root cause is GC doesn't wait for all the informers to be synced.
>Here is a series of events that lead to the missing deletion event of the RC:
>
>GC starts the informer for Pods, and sees the pod before it sees the RC, so it creates a virtual node for the RC, and push the RC in the attempToDelete
>In attemptToDeleteWorker(), GC checks if the RC is still there and it sees it's still there, so it doesn't delete the virtual node. Thus not deleting the Pod.
>Later, the test deletes the RC. After that, the informer for RC starts to list and watch, and the RC is already deleted so the reflector never reports the deletion of the RC.
>The solution is wait for all the informers HasSynced() before running any GC workers.

In the case of the missed RC delete event, doesn't that leave a virtual node for the RC which is never re-verified with an existence check? Even if there was a future update to the owned pod, the virtual node which owns the pod doesn't seem to be re-queued for a deletion attempt (correct me if I'm wrong).

I'm not sure that syncing informer checks would ever totally close this gap: it should be assumed that events can be missed. Shouldn't the entire graph be periodically reconciled somehow to ensure consistency? For example, at the least, periodically attempting to delete any virtual nodes that exists in order to account for the inevitable missed event.

Given all that, I still think your suggestion would further minimize the possibility of getting into the problematic states in the first place. There would be some extra refactoring to introduce an independent internal lifecycle to the workers.